librtt/librtt.c

```diff
@@ -0,0 +1,432 @@
+ * the RTT remote side e.g. openocd to find descriptor location during memory
+ * scan, so as not to mislead the descriptor scan procedure, this tag needs
+ * to be simply hidden before it is copied into RAM together with initialized
+ * descriptors, that's why it is written backwards.
+ * - Default LIBRTT_TAG_REVERSED="EPIP TTR", which corresponds to "RTT PIPE"
+ */
+
+/* clang-format off */
+
+#ifndef LIBRTT_TAG_BACKWARD
+#define LIBRTT_TAG_BACKWARD     "EPIP TTR"
+#endif
+
+
+#ifndef LIBRTT_TXCHANNELS
+#define LIBRTT_TXCHANNELS       2
+#define LIBRTT_TXCHANNELS_NAMES { "rtt1_tx", "rtt2_tx" }
+#endif
+
+#ifndef LIBRTT_RXCHANNELS
+#define LIBRTT_RXCHANNELS       2
+#define LIBRTT_RXCHANNELS_NAMES { "rtt1_rx", "rtt2_rx" }
+#endif
+
+/* clang-format on */
+
+
+struct rtt_pipe {
+	const char *name;
+	volatile unsigned char *ptr;
+	unsigned int sz;
+	volatile unsigned int wr;
+	volatile unsigned int rd;
+	unsigned int flags;
+};
+
+
+struct rtt_desc {
+	char tag[16];
+	unsigned int txChannels;
+	unsigned int rxChannels;
+	struct rtt_pipe channel[LIBRTT_TXCHANNELS + LIBRTT_RXCHANNELS];
+};
+
+
+static const char librtt_tagReversed[] = LIBRTT_TAG_BACKWARD;
+static const size_t librtt_tagLength = sizeof(librtt_tagReversed) - 1;
+static const char *const librtt_txName[LIBRTT_TXCHANNELS] = LIBRTT_TXCHANNELS_NAMES;
+static const char *const librtt_rxName[LIBRTT_RXCHANNELS] = LIBRTT_RXCHANNELS_NAMES;
+
+
+static struct {
+	volatile struct rtt_desc *rtt;
+	librtt_cacheOp_t invalFn;
+	librtt_cacheOp_t cleanFn;
+	unsigned int lastRd[LIBRTT_TXCHANNELS];
+} librtt_common = { 0 };
+
+
+static void performCacheOp(librtt_cacheOp_t op, volatile unsigned char *buf, unsigned int sz, unsigned int rd, unsigned int wr)
+{
+	if ((rd == wr) || (op == NULL)) {
+		return;
+	}
+
+	if (wr < rd) {
+		op((void *)(buf + rd), sz - rd);
+		op((void *)buf, wr + 1);
+	}
+	else {
+		op((void *)(buf + rd), wr - rd + 1);
+	}
+}
+
+
+int librtt_checkTx(unsigned int ch)
+{
+	if ((librtt_common.rtt == NULL) || (ch >= librtt_common.rtt->txChannels)) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+
+int librtt_checkRx(unsigned int ch)
+{
+	if ((librtt_common.rtt == NULL) || (ch >= librtt_common.rtt->rxChannels)) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+
+ssize_t librtt_read(unsigned int ch, void *buf, size_t count)
+{
+	volatile struct rtt_desc *rtt = librtt_common.rtt;
+
+	if (librtt_checkRx(ch) < 0) {
+		return -ENODEV;
+	}
+
+	dataMemoryBarrier();
+	ch += rtt->txChannels;
+	volatile unsigned char *srcBuf = rtt->channel[ch].ptr;
+	unsigned char *dstBuf = (unsigned char *)buf;
+	unsigned int sz = rtt->channel[ch].sz - 1;
+	unsigned int rd = rtt->channel[ch].rd & sz;
+	unsigned int wr = rtt->channel[ch].wr & sz;
+	size_t todo = count;
+
+	performCacheOp(librtt_common.invalFn, (void *)srcBuf, sz + 1, rd, wr);
+	while ((todo != 0) && (rd != wr)) {
+		*dstBuf++ = srcBuf[rd];
+		rd = (rd + 1) & sz;
+		todo--;
+	}
+
+	rtt->channel[ch].rd = rd;
+
+	dataMemoryBarrier();
```

Comment:
This should actually be before the write to `rd` - if reorder happens and `rd` is set first, host/jlink may overwrite data before librtt has a chance to read it.